rust/lance-io/src/object_store/providers/aws.rs

```diff
@@ -269,6 +283,58 @@ pub async fn build_aws_credential(
     }
 }
 
+async fn build_aws_credential_with_storage_options_provider(
+    storage_options_provider: Arc<dyn StorageOptionsProvider>,
+    credentials_refresh_offset: Duration,
+    credentials: Option<AwsCredentialProvider>,
+    storage_options_credentials: Option<StaticCredentialProvider<ObjectStoreAwsCredential>>,
+    expires_at_millis: Option<u64>,
+) -> Result<AwsCredentialProvider> {
+    match (expires_at_millis, credentials, storage_options_credentials) {
+        // Case 1: provider + credentials + expiration time
+        (Some(expires_at), Some(cred), _) => {
+            Ok(Arc::new(
+                DynamicStorageOptionsCredentialProvider::new_with_initial_credential(
+                    storage_options_provider,
+                    credentials_refresh_offset,
+                    cred.get_credential().await?,
+                    expires_at,
+                ),
+            ))
+        }
+        // Case 2: provider + storage_options (with valid credentials) + expiration time
+        (Some(expires_at), None, Some(cred)) => {
+            Ok(Arc::new(
+                DynamicStorageOptionsCredentialProvider::new_with_initial_credential(
+                    storage_options_provider,
+                    credentials_refresh_offset,
+                    cred.get_credential().await?,
+                    expires_at,
+                ),
+            ))
+        }
+        // Case 3: provider + storage_options without expiration - FAIL
+        (None, None, Some(_)) => Err(Error::IO {
+            source: Box::new(std::io::Error::other(
+                "expires_at_millis is required when using storage_options_provider with storage_options",
+            )),
+            location: location!(),
+        }),
+        // Case 4: provider + credentials without expiration - FAIL
+        (None, Some(_), _) => Err(Error::IO {
+            source: Box::new(std::io::Error::other(
+                "expires_at_millis is required when using storage_options_provider with credentials",
+            )),
+            location: location!(),
+        }),
```

Comment:
I think we can trust if they pass a credential provider they already handled expiration, right?